dup runtime test fails

The dup runtime test fails with:

```
$ make TARGET=dup run                                                                                                                                                                                   
make[1]: Entering directory '/share/src/nanovms/master01/contgen'                                                                                                                                                                              
make[1]: Nothing to be done for 'all'.                                                                                                                                                                                                         
make[1]: Leaving directory '/share/src/nanovms/master01/contgen'                                                                                                                                                                               
make[1]: Entering directory '/share/src/nanovms/master01/mkfs'                                                                                                                                                                                 
make[1]: Entering directory '/share/src/nanovms/master01/boot'                                                                                                                                                                                 
make[1]: Entering directory '/share/src/nanovms/master01/stage3'                                                                                                                                                                               
make[1]: Entering directory '/share/src/nanovms/master01/test/runtime'                                                                                                                                                                         
make[1]: Nothing to be done for 'all'.                                                                                                                                                                                                         
make[1]: Leaving directory '/share/src/nanovms/master01/boot'                                                                                                                                                                                  
make[1]: Nothing to be done for 'all'.                                                                                                                                                                                                         
make[1]: Leaving directory '/share/src/nanovms/master01/mkfs'                                                                                                                                                                                  
make[1]: Nothing to be done for 'all'.                                                                                                                                                                                                         
make[1]: Leaving directory '/share/src/nanovms/master01/stage3'                                                                                                                                                                                
make[1]: Nothing to be done for 'dup'.                                                                                                                                                                                                         
make[1]: Leaving directory '/share/src/nanovms/master01/test/runtime'                                                                                                                                                                          
MKFS    image                                                                                                                                                                                                                                  
metadata (arguments:(0:dup) children:(kernel:() dup:()) program:/dup fault:t environment:())                                                                                                                                                   
qemu-system-x86_64 -m 2G -display none -serial stdio -drive if=none,id=hd0,format=raw,file=/share/src/nanovms/master01/output/image/disk.raw -device virtio-scsi-pci,id=scsi0 -device scsi-hd,bus=scsi0.0,drive=hd0 -device isa-debug-exit -no\
-reboot  -device virtio-net,netdev=n0 -netdev user,id=n0,hostfwd=tcp::8080-:8080,hostfwd=tcp::9090-:9090,hostfwd=udp::5309-:5309 -enable-kvm -cpu host || exit $(($?>>1))                                                                      
assigned: 10.0.2.15                                                                                                                                                                                                                            
Error: dup2(2, i) == i -- failed at /share/src/nanovms/master01/test/runtime/dup.c:34                                                                                                                                                          
Makefile:132: recipe for target 'run' failed                                                                                                                                                                                                   
make: *** [run] Error 1  
$
```
... and it also fails when running on the host, albeit differently:

```
$ output/test/runtime/bin/dup                                                                                                                                                                           
Error: dup(2) == (3 + i) -- failed at /share/src/nanovms/master01/test/runtime/dup.c:21
$
```

Need to determine if this is just the test, the dup implementation or both - and fix.

--- a/Makefile
+++ b/Makefile
@@ -89,7 +89,7 @@ test test-noaccel: mkfs boot stage3
 	$(Q) $(MAKE) -C test test
 	$(Q) $(MAKE) runtime-tests$(subst test,,$@)
 
-RUNTIME_TESTS=	aio creat epoll eventfd fallocate fcntl fst getdents getrandom hw hws mkdir mmap pipe readv rename sendfile signal socketpair time unlink thread_test vsyscall write writev
+RUNTIME_TESTS=	aio creat dup epoll eventfd fallocate fcntl fst getdents getrandom hw hws mkdir mmap pipe readv rename sendfile signal socketpair time unlink thread_test vsyscall write writev
 
 .PHONY: runtime-tests runtime-tests-noaccel
 

--- a/src/runtime/heap/id.c
+++ b/src/runtime/heap/id.c
@@ -341,3 +341,14 @@ id_heap create_id_heap_backed(heap meta, heap map, heap parent, bytes pagesize)
     }
     return i;
 }
+
+void id_heap_set_next(id_heap h, u64 next)
+{
+    rangemap_foreach(h->ranges, n) {
+        id_range r = (id_range)n;
+        if (point_in_range(r->n.r, next))
+            r->next_bit = next;
+        else if (r->n.r.start > next)
+            r->next_bit = 0;
+    }
+}

--- a/src/runtime/heap/id.h
+++ b/src/runtime/heap/id.h
@@ -23,7 +23,13 @@ id_heap allocate_id_heap(heap meta, heap map, bytes pagesize); /* id heap with n
 #define id_heap_set_randomize(__h, __r) ((__h)->set_randomize(__h, __r))
 #define id_heap_alloc_subrange(__h, __c, __s, __e) ((__h)->alloc_subrange(__h, __c, __s, __e))
 
+/* Provides a hint as to what id should be allocated next. */
+void id_heap_set_next(id_heap h, u64 next);
+
+/* If count == 1, the return value is guaranteed to be the lowest-numbered
+ * non-allocated id starting from min. */
 static inline u64 id_heap_alloc_gte(id_heap h, bytes count, u64 min)
 {
+    id_heap_set_next(h, min);
     return id_heap_alloc_subrange(h, count, min, infinity);
 }

--- a/test/runtime/fcntl.c
+++ b/test/runtime/fcntl.c
@@ -5,6 +5,13 @@
 #include <fcntl.h>
 #include <errno.h>
 
+#define test_assert(expr) do { \
+    if (!(expr)) { \
+        printf("Error: %s -- failed at %s:%d\n", #expr, __FILE__, __LINE__); \
+        exit(EXIT_FAILURE); \
+    } \
+} while (0)
+
 /* covers F_GETLK, F_SETLK, F_SETLKW; expect = 0 for success, errno otherwise */
 void test_lk(int fd, int cmd, struct flock *lock, int expect)
 {
@@ -30,6 +37,24 @@ void test_lk(int fd, int cmd, struct flock *lock, int expect)
     exit(EXIT_FAILURE);
 }
 
+void test_dupfd(int fd)
+{
+    int new_fd1, new_fd2;
+
+    new_fd1 = fcntl(fd, F_DUPFD, 0);
+    test_assert(new_fd1 > fd);
+    test_assert(close(new_fd1) == 0);
+
+    new_fd1 = fcntl(fd, F_DUPFD, 10);
+    test_assert(new_fd1 == 10);
+
+    new_fd2 = fcntl(new_fd1, F_DUPFD, 0);
+    test_assert(new_fd2 < new_fd1);
+
+    test_assert(close(new_fd1) == 0);
+    test_assert(close(new_fd2) == 0);
+}
+
 int main(int argc, char **argv)
 {
     struct flock lock;
@@ -43,6 +68,8 @@ int main(int argc, char **argv)
     test_lk(fd, F_GETLK,  &lock, 0);
     test_lk(fd, F_SETLK,  &lock, 0);
     test_lk(fd, F_SETLKW, &lock, 0);
+
+    test_dupfd(fd);
 
     printf("test passed\n");
     return EXIT_SUCCESS;

